Consistently use TAI instead of issuer_id (#91)
davidben/merkle-tree-certs: draft-davidben-tls-merkle-tree-certs.md

## Identifying CAs and Batches {#identifying}

A Merkle Tree CA's `issuer_id` is a trust anchor identifier, defined in {{Section 3 of !I-D.beck-tls-trust-anchor-ids}}. 
However, unlike an X.509 CA, the entire OID arc rooted at the identifier is associated with the CA. 
OIDs under this arc are used to identify batches below.

An individual batch from a Merkle Tree CA also has an associated trust anchor identifier. It is determined by appending the batch number to the CA's `issuer_id`.

> In the following it is called `batch_id`.

Reviewer: (Looks like you already did this one.)

A Merkle Tree CA's `issuer_id` is a trust anchor identifier, defined in {{Section 3 of !I-D.beck-tls-trust-anchor-ids}}. However, unlike an X.509 CA, the entire OID arc rooted at the identifier is associated with the CA. OIDs under this arc are used to identify batches below.

> An individual batch from a Merkle Tree CA also has an associated trust anchor identifier. It is determined by appending the batch number to the CA's `issuer_id`.
> In the following it is called `batch_id`.

Reviewer: ```suggestion
An individual batch from a Merkle Tree CA also has an associated trust anchor identifier, called a `batch_id`. It is determined by appending the batch number, as an OID component, to the CA's `issuer_id`.
```

Reviewer: applied.

struct {
    uint8 distinguisher = 2;

>     TrustAnchorIdentifier trust_anchor;

Reviewer: Oops, nice catch! Thanks! This was just a mistake in rebasing over TAI.

Looks like the text down in line 562 needs to be updated since they refer to the structures. Same with line 627. It's a bit awkward because {{parameters}} and {{identifying}} still call it an `issuer_id`, but then the structures all say `trust_anchor`. Probably that should be renamed to be consistent across the board.

(Do you mind fixing that in this PR? If not, I can merge this and then finish it up.)

Author: While applying the changes you proposed, I noticed that we might also remove the batch number from the `Hash...` structs altogether, and require the trust anchor to identify the specific batch. See [0e5eeca](https://github.com/davidben/merkle-tree-certs/pull/91/commits/0e5eecad5dc6d2cdbfa24b1ef7f47c60c01a2187). Let me know if you like the suggestion.

Independent of that, I'm a bit worried that it became easy to confuse when to use the CA specific TAI and when a batch specific TAI is required, i.e., if the batch number is included in the TAI. Maybe it is worth making this a bit clearer in a later change. I'm not sure if it is actually required, though, and how to do this discrimination elegantly.

Reviewer: > Independent of that, I'm a bit worried that it became easy to confuse when to use the CA specific TAI and when a batch specific TAI is required, i.e., if the batch number is included in the TAI. Maybe it is worth making this a bit clearer in a later change. I'm not sure if it is actually required, though, and how to do this discrimination elegantly.

Hmm, that is a good point. One thought: what if we continue to call the CA-specific ID the `issuer_id` or perhaps the `ca_id` and then the batch-specific ID is called the `batch_id`. Since `BikeshedCertificate` is meant to be extensible, that probably needs to say `trust_anchor` with prose saying something like:

> A proof's `trust_anchor` field is a trust anchor identifier (see {{Section 3 of !I-D.beck-tls-trust-anchor-ids}} and {{Section 4.1 of !I-D.beck-tls-trust-anchor-ids}}), which determines the proof's type and issuer. It is analogous to an X.509 trust anchor's subject name. When the issuer is a Merkle Tree CA, the `trust_anchor` is a batch's `batch_id`, as described in {{identifying}}.

And then all the MTC-specific structures can say `issuer_id` or `batch_id` depending on which it's meant to be. WDYT?

Author: Sounds reasonable to me. I updated the PR to include `issuer_id` and `batch_id` where appropriate. I hope it is consistent.

struct {
    uint8 distinguisher = 2;
    TrustAnchorIdentifier trust_anchor;
    uint64 index;
    AbridgedAssertion abridged_assertion;
} HashAssertionInput;
~~~

> The `trust_anchor` is set to the batch specific trust anchor identifier, i.e., the conncatelation of the CA's `trust_anchor` with the batch_number.

Reviewer: ```suggestion
The `trust_anchor` is set to the batch-specific trust anchor identifier, i.e. with the batch number appended as described in {{identifying}}.
```

(Just to use the same wording.)

} BikeshedCertificate;
~~~

A proof's `trust_anchor` field is a trust anchor identifier (see {{Section 3 of !I-D.beck-tls-trust-anchor-ids}} and {{Section 4.1 of !I-D.beck-tls-trust-anchor-ids}}), which determines the proof's type and issuer.
It is analogous to an X.509 trust anchor's subject name.

> When the issuer is a Merkle Tree CA, the `trust_anchor` is a batch's batch_id, as described in {{identifying}}.

Reviewer: ```suggestion
When the issuer is a Merkle Tree CA, the `trust_anchor` is a batch's `batch_id`, as described in {{identifying}}.
```
